### PR TITLE
enhancement(titles): improved parsing

### PR DESCRIPTION
**Pull request by s0up4200**
Output of a mix of normal and troublesome titles: (From commit https://github.com/autobrr/omegabrr/pull/37/commits/64d572989d3161fca09c2705cd919641cd84838f)

```
30?Rock
Aaahh!!!?Real?Monsters,Aaahh*Real?Monsters
All?or?Nothing:?Arsenal,All?or?Nothing*Arsenal
American?Dad!,American?Dad
And?Just?Like?That...,And?Just?Like?That
Arthur!?And?the?Square?Knights?of?the?Round?Table,Arthur*And?the?Square?Knights?of?the?Round?Table
Be?Cool??Scooby?Doo!,Be?Cool?Scooby?Doo
Bob's?Burgers,Bobs?Burgers
Brooklyn?Nine?Nine
Chicago?P.D.,Chicago?P?D
Darkness:?Those?Who?Kill,Darkness*Those?Who?Kill
Harry?and?Meghan,Harry?&?Meghan
Hello?Tomorrow!,Hello?Tomorrow
Marvel's?The?Punisher,Marvels?The?Punisher
MasterChef?(US),MasterChef*US,MasterChef
Mayans?M.C.,Mayans?M?C
Mr.?Robot,Mr??Robot,Mr?Robot,Mr*Robot
S.W.A.T.,S?W?A?T
Scooby?Doo*Mystery?Incorporated,Scooby?Doo!?Mystery?Incorporated
Slow?Horses
The?Handmaid's?Tale,The?Handmaids?Tale
The?Last?of?Us
Tom?Clancy's?Jack?Ryan,Tom?Clancys?Jack?Ryan
Whose?Line?is?it?Anyway?,Whose?Line?is?it?Anyway
Whose?Line?Is?It?Anyway??(US),Whose?Line?Is?It?Anyway*US,Whose?Line?Is?It?Anyway
Your?Honor,Your?Honor?(US),Your?Honor*US
```

### CODE REVIEW

**Comment by s0up4200** [comment]
A mix of normal and troublesome titles:

```
The?Last?of?Us

Hello?Tomorrow!,Hello?Tomorrow

Arthur!?And?the?Square?Knights?of?the?Round?Table,Arthur?And?the?Square?Knights?of?the?Round?Table,Arthur*And?the?Square?Knights?of?the?Round?Table

Aaahh!!!?Real?Monsters,Aaahh?Real?Monsters,Aaahh*Real?Monsters

Scooby-Doo?Mystery?Incorporated,Scooby?Doo*Mystery?Incorporated,Scooby-Doo!?Mystery?Incorporated,Scooby?Doo!?Mystery?Incorporated

All?or?Nothing:?Arsenal,All?or?Nothing?Arsenal,All?or?Nothing*Arsenal

Bob's?Burgers,Bobs?Burgers

Brooklyn?Nine-Nine,Brooklyn?Nine?Nine

Chicago?P?D,Chicago?P.D.

Whose?Line?is?it?Anyway?,Whose?Line?is?it?Anyway

Whose?Line?Is?It?Anyway??(US),Whose?Line?Is?It?Anyway*

Mr?Robot,Mr.?Robot,Mr??Robot

Darkness:?Those?Who?Kill,Darkness?Those?Who?Kill,Darkness*Those?Who?Kill

The?Handmaid's?Tale,The?Handmaids?Tale

Harry?&?Meghan,Harry?and?Meghan

Marvel's?The?Punisher,Marvels?The?Punisher

Mayans?M.C.,Mayans?M?C

MasterChef*,MasterChef?(US)

S.W.A.T.,S?W?A?T
```

**Comment by s0up4200** [comment]
Simplified the code.

```
Tom?Clancy's?Jack?Ryan,Tom?Clancys?Jack?Ryan
The?Last?of?Us
Aaahh!!!?Real?Monsters,Aaahh*Real?Monsters
MasterChef?(US),MasterChef*US
Whose?Line?Is?It?Anyway?*US,Whose?Line?Is?It?Anyway*(US),Whose?Line?Is?It?Anyway??(US)
Whose?Line?is?it?Anyway?
Hello?Tomorrow!,Hello?Tomorrow
Be?Cool?Scooby?Doo!,Be?Cool?Scooby?Doo
Scooby-Doo!?Mystery?Incorporated,Scooby?Doo*Mystery?Incorporated
All?or?Nothing:?Arsenal,All?or?Nothing*Arsenal
American?Dad,American?Dad!,And?Just?Like?That,And?Just?Like?That...
Arthur!?And?the?Square?Knights?of?the?Round?Table,Arthur*And?the?Square?Knights?of?the?Round?Table
Bobs?Burgers,Bob's?Burgers
Brooklyn?Nine?Nine
Chicago?P.D.,Chicago?P?D
Darkness*Those?Who?Kill,Darkness:?Those?Who?Kill
Harry?&?Meghan,Harry?and?Meghan
Marvel's?The?Punisher,Marvels?The?Punisher
Mayans?M.C.,Mayans?M?C
Mr.?Robot,Mr??Robot,Mr?RobotMr*Robot
S.W.A.T.,S?W?A?T
The?Handmaids?Tale,The?Handmaid's?Tale
```

File: internal/processor/title.go [hunk 1]
```diff
@@ -32,39 +29,31 @@ func processTitle(title string, matchRelease bool) []string {
...
-		t.Add(replace, matchRelease)
-
-		split := strings.Split(title, ":")
-		if len(split) > 1 {
-			first := strings.ReplaceAll(split[0], " ", "?")
-			second := strings.ReplaceAll(strings.Trim(split[1], " "), " ", "?")
-			part := fmt.Sprintf("%v*%v", first, second)
-
-			t.Add(part, matchRelease)
+	if re, err := regexp.Compile(`\s\(([A-Za-z]{2})\)$`); err == nil {
```
Comment (nuxencs) [inline internal/processor/title.go:32]: from here to line 58 could probably be replaced by this, but would need further testing:
```
if regexp.MustCompile(`[[:punct:]]`).MatchString(title) {
		regexQuestionmark := regexp.MustCompile(`[?]{2,}`)
		regexRemove := regexp.MustCompile(`[^[:alnum:]]`)

		title = strings.TrimRight(title, ".!")
		title = strings.ReplaceAll(title, "'", "")
		title = strings.ReplaceAll(title, "’", "")
		title = strings.ReplaceAll(title, ",", "")
		title = strings.ReplaceAll(title, "&", "and")

		replace := regexRemove.ReplaceAllString(title, "?")
		replace = regexQuestionmark.ReplaceAllString(replace, "*")
		replace = strings.TrimRight(replace, "?*")
		t.Add(replace, matchRelease)
	}
```

**Review by nuxencs — COMMENTED** [review]
left a comment on line 32 in `title.go`, but can't comment on the dependencies. guess you just removed unneeded ones 😄

File: internal/processor/title.go [hunk 2]
```diff
@@ -18,42 +18,55 @@ func processTitle(title string, matchRelease bool) []string {
...
+		title = strings.ReplaceAll(title, "’", "")
 		title = strings.ReplaceAll(title, ",", "")
-		title = strings.ReplaceAll(title, "?*", "*")
 		title = strings.ReplaceAll(title, "&", "and")
-		title = strings.ReplaceAll(title, " ", "?")
+
+		// Replace all non-alphanumeric characters with "?", and all occurrences of "???" with "*"
+		replace := regexReplace.ReplaceAllString(title, "?")
+		replace = regexQuestionmark.ReplaceAllString(replace, "*")
+		t.Add(strings.TrimRight(replace, "?*"), matchRelease)
```
Comment (nuxencs) [inline internal/processor/title.go:50]: should probably add ` ` to the TrimRight so it looks like this:
`t.Add(strings.TrimRight(replace, "?* "), matchRelease)`

**Review by zze0s — APPROVED** [review]
Great improvement! Well done @s0up4200 @nuxencs

**Pull request merged**